### PR TITLE
[RFC] Enable -Wconversion: tag.c

### PR DESCRIPTION
**Pull request by vheon**
I tried to tackle this one. The only doubt that I have is for lines that interact with `num_matches` (https://github.com/neovim/neovim/pull/2803/files#diff-bb67c9d81d07a0359baaea74957f649dR1968) and `mincount` (https://github.com/neovim/neovim/pull/2803/files#diff-bb67c9d81d07a0359baaea74957f649dR1905): I wanted to make them `size_t` as well but they got passed around to other functions from the outside and following their uses wasn't easy, so I would like to hear some other opinion about those.

### CODE REVIEW

**Comment by justinmk** [comment]
The build failures are legitimate. Let us know if you have any questions.

**Comment by vheon** [comment]
> The build failures are legitimate.

I got the same failure on master.

>  Let us know if you have any questions.

The only question that I have is if it is enough the casting to int done for `match_count`.

File: src/nvim/tag.c [hunk 1]
```diff
@@ -1409,7 +1409,7 @@ find_tags (
           conv_line = string_convert(&vimconv, lbuf, NULL);
           if (conv_line != NULL) {
             /* Copy or swap lbuf and conv_line. */
-            len = (int)STRLEN(conv_line) + 1;
+            len = STRLEN(conv_line) + 1;
```
Comment (splinterofchaos) [inline src/nvim/tag.c]: Looks like the declaration of `len` can be moved here.

**Comment by splinterofchaos** [comment]
> > The build failures are legitimate.

> I got the same failure on master.

Well, Travis says for the 32-bit build:
```
/home/travis/build/neovim/neovim/src/nvim/tag.c: In function ‘find_tags’:
/home/travis/build/neovim/neovim/src/nvim/tag.c:1644:13: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
/home/travis/build/neovim/neovim/src/nvim/tag.c:1653:13: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
/home/travis/build/neovim/neovim/src/nvim/tag.c:1662:11: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
/home/travis/build/neovim/neovim/src/nvim/tag.c:1697:13: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
/home/travis/build/neovim/neovim/src/nvim/tag.c:1762:19: error: conversion to ‘char_u’ from ‘int’ may alter its value [-Werror=conversion]
/home/travis/build/neovim/neovim/src/nvim/tag.c:1764:19: error: conversion to ‘char_u’ from ‘int’ may alter its value [-Werror=conversion]
/home/travis/build/neovim/neovim/src/nvim/tag.c: In function ‘expand_tags’:
/home/travis/build/neovim/neovim/src/nvim/tag.c:2714:35: error: conversion to ‘char_u’ from ‘int’ may alter its value [-Werror=conversion]
/home/travis/build/neovim/neovim/src/nvim/tag.c:2716:56: error: conversion to ‘size_t’ from ‘int’ may change the sign of the result [-Werror=sign-conversion]
```

And clang:
```
/home/travis/build/neovim/neovim/src/nvim/tag.c:2716:56: error: implicit
      conversion changes signedness: 'long' to 'unsigned long'
      [-Werror,-Wsign-conversion]
      memmove((*file)[i] + c, t_p.fname, t_p.fname_end - t_p.fname);
      ~~~~~~~                            ~~~~~~~~~~~~~~^~~~~~~~~~~
1 error generated.
```
Please fix these.

**Comment by vheon** [comment]
@splinterofchaos I addressed most of what you pointed out to me. Three of them are still there though and I'm a bit puzzled by them:

```
/home/travis/build/vheon/neovim/src/nvim/tag.c: In function ‘find_tags’:
/home/travis/build/vheon/neovim/src/nvim/tag.c:1753:19: error: conversion to ‘char_u’ from ‘int’ may alter its value [-Werror=conversion]
/home/travis/build/vheon/neovim/src/nvim/tag.c:1755:19: error: conversion to ‘char_u’ from ‘int’ may alter its value [-Werror=conversion]
/home/travis/build/vheon/neovim/src/nvim/tag.c: In function ‘expand_tags’:
/home/travis/build/vheon/neovim/src/nvim/tag.c:2705:35: error: conversion to ‘char_u’ from ‘int’ may alter its value [-Werror=conversion]
```

The first two are from this snippet of code https://github.com/neovim/neovim/pull/2803/files#diff-bb67c9d81d07a0359baaea74957f649dR1753, (I'll report it down for convenience)

```C
            if (orgpat.regmatch.rm_ic && !match_no_ic)
              mtt += MT_IC_OFF;
            if (match_re)
              mtt += MT_RE_OFF;
```

mtt is a `char_u` while `MT_IC_OFF` and `MT_RE_OFF` are `#define` for 4 and 8 respectively. Unless I misunderstood, since we're adding something to `mtt` gcc apparently conver it to an `int` (???). So how should cast appropriately?

**Comment by splinterofchaos** [comment]
Actually, I don't understand why you made `mtt` a `char`. Looks like we use it as an `int`, or maybe better: `unsigned int`.

Not sure why you'd get that error though.

**Comment by vheon** [comment]
@splinterofchaos I made it a `char_u` because it never gets greater then 16, and in one point its value is inserted as a char in a string; so I thought that was right to make it a `char_u`. Being said that I could easily change it to an `unsigned int` and cast it to `char_u` only in the case it gets inserted into the string.

**Comment by vheon** [comment]
@splinterofchaos I've made the changes you suggested.

**Comment by splinterofchaos** [comment]
LGTM

Those checks for `>= 0` were all `assert`s, right? Just want to make sure because if they were conditions, that would indicate that the value could become negative, but having a bunch of `assert(x >= 0)` is pointless if we can make it unsigned.

**Comment by vheon** [comment]
@splinterofchaos yes, they were all assert for `>= 0`. I didn't catch them at first because I compiled only with clang on my machine and I didn't activate travis on my fork.

File: src/nvim/tag.c [hunk 2]
```diff
@@ -2624,7 +2612,7 @@ static char_u *expand_tag_fname(char_u *fname, char_u *tag_fname, int expand)
  */
 static int test_for_current(char_u *fname, char_u *fname_end, char_u *tag_fname, char_u *buf_ffname)
 {
-  int c;
+  char_u c;
```
Comment (justinmk) [inline src/nvim/tag.c]: why not `char`?

File: src/nvim/tag.c [hunk 3]
```diff
@@ -2711,10 +2699,10 @@ expand_tags (
       c = (int)(t_p.tagname_end - t_p.tagname);
```
Comment (justinmk) [inline src/nvim/tag.c]: here `int` is assigned to `char_u`, that does not seem right.

**Comment by justinmk** [comment]
@vheon could you rebase this? It might require some style fixes, our client check is a bit more strict now.

**Comment by cacplate** [comment]
@justinmk this can be closed since I adopted it.

**Pull request closed without merging**